### PR TITLE
Create and delete support for block Metro volume

### PR DESCRIPTION
**Pull request by santhoshatdell**
# Description
- Create and delete support for block Metro volume. This feature is only implemented at the driver level like PowerMax Metro, without the involvement of replication module.
- Two backend volumes get created for single PV upon replication request specified in Storage class. If the _ReclaimPolicy_ is 'Retain', both backend volumes along with PV are retained. If it is 'Delete', all are deleted.
- Updates to PV's volumeHandle to contain both backend volumes ID will be in next [PR](https://github.com/dell/csi-powerstore/pull/345).

# GitHub Issues
List the GitHub issues impacted by this PR:

| GitHub Issue # |
| -------------- |
| https://github.com/dell/csm/issues/1443 |

# Checklist:

- [x] I have performed a self-review of my own code to ensure there are no formatting, vetting, linting, or security issues
- [x] I have verified that new and existing unit tests pass locally with my changes
- [x] I have not allowed coverage numbers to degenerate (controller pkg 42.0% to 42.3%. Luke's PR will include changes to report better code coverage number)
- [ ] I have maintained at least 90% code coverage
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Backward compatibility is not broken

# How Has This Been Tested?

- Tested that Metro PV is created and then deleted successfully. Remote backed volume is also created and deleted on the remote array as subsequent calls are made on source PV. (Logs are added to the ticket 23538)
- Tested that configure metro call is idempotent, also tested the failure scenarios as added in the code.
- Added new unit tests.
- Verified that non-metro PV creation and deletion continues to work. Updated _tests/simple/simple.yaml_ to reflect CSM [doc](https://dell.github.io/csm-docs/docs/csidriver/test/powerstore/).

### CODE REVIEW

**Comment by santhoshatdell** [comment]
It looks GitHub checks are failing due to go 1.22 usage in common-github-actions (I'll try to update it there), but please review the code changes.

File: pkg/controller/controller.go [hunk 1]
```diff
@@ -374,8 +381,22 @@ func (s *Service) CreateVolume(ctx context.Context, req *csi.CreateVolumeRequest
...
+		metroSession, err := arr.GetClient().ConfigureMetroVolume(ctx, volID, &gopowerstore.MetroConfig{
+			RemoteSystemID: remoteSystem.ID,
+		})
+		if err != nil {
+			if apiError, ok := err.(gopowerstore.APIError); ok && apiError.ReplicationSessionAlreadyCreated() { // idempotency check
+				log.Debugf("Metro has already been configured on volume %s", volID)
+			} else {
+				return nil, status.Errorf(codes.Internal, "can't configure metro on volume: %s", err.Error())
+			}
+		} else {
```
Comment (lukeatdell) [inline pkg/controller/controller.go:397]: could remove the `else` if you wanted to, here. Let me know and I can make the change so we don't conflict.

**Review by lukeatdell — APPROVED** [review]

File: pkg/controller/controller.go [hunk 2]
```diff
@@ -538,6 +559,25 @@ func (s *Service) DeleteVolume(ctx context.Context, req *csi.DeleteVolumeRequest
 				"unable to delete volume -- %d snapshots based on this volume still exist.", len(listSnaps))
 		}
 
+		// Check if volume has metro session and end it
+		volume, err := arr.GetClient().GetVolume(ctx, id)
+		if err != nil {
+			if apiError, ok := err.(gopowerstore.APIError); ok && apiError.NotFound() {
+				log.Infof("Volume %s not found, it may have been deleted.", id)
+				return &csi.DeleteVolumeResponse{}, nil
+			}
```
Comment (donatwork) [inline pkg/controller/controller.go:568]: Would a warning be better? You don't have to change the code either way.

**Review by donatwork — APPROVED** [review]

File: pkg/controller/controller.go (hunk 2 above)
Comment (santhoshatdell) [reply pkg/controller/controller.go:568]: I noticed repeated delete calls to the driver whether it is metro or regular volume. In those cases, 'info' seemed fine.

**Pull request merged**